src/scenic/core/regions.py

```diff
@@ -281,883 +306,2816 @@ def __hash__(self):
+        super().__init__(name, *additionalDeps, orientation=orientation)
+
+        self.z = z
+
+        # If our polygons aren't defined yet, then wait till later
+        if isLazy(self):
+            return
+
+        if polygon is None and points is None:
+            raise ValueError('must specify points or polygon for PolygonalRegion')
+        if polygon is None:
+            points = tuple(pt[:2] for pt in points)
+            if len(points) == 0:
+                raise ValueError('tried to create PolygonalRegion from empty point list!')
+            for point in points:
+                if any(needsSampling(coord) for coord in point):
+                    raise ValueError('only fixed PolygonalRegions are supported')
+            self.points = points
+            polygon = shapely.geometry.Polygon(points)
+
+        if isinstance(polygon, shapely.geometry.Polygon):
+            self.polygons = shapely.geometry.MultiPolygon([polygon])
+        elif isinstance(polygon, shapely.geometry.MultiPolygon):
+            self.polygons = polygon
+        else:
+            raise ValueError(f'tried to create PolygonalRegion from non-polygon {polygon}')
+        if not self.polygons.is_valid:
+            raise ValueError('tried to create PolygonalRegion with '
+                             f'invalid polygon {self.polygons}')
+
+        if (points is None and len(self.polygons.geoms) == 1
+            and len(self.polygons.geoms[0].interiors) == 0):
+            self.points = tuple(self.polygons.geoms[0].exterior.coords[:-1])
+
+        if self.polygons.is_empty:
+            raise ValueError('tried to create empty PolygonalRegion')
+
+        triangles = []
+        for polygon in self.polygons.geoms:
+            triangles.extend(triangulatePolygon(polygon))
+        assert len(triangles) > 0, self.polygons
+        self.trianglesAndBounds = tuple((tri, tri.bounds) for tri in triangles)
+        areas = (triangle.area for triangle in triangles)
+        self.cumulativeTriangleAreas = tuple(itertools.accumulate(areas))
+
+    @cached_property
+    def footprint(self):
+        return PolygonalFootprintRegion(self.polygons)
+
+    def containsPoint(self, point):
+        return self.footprint.containsPoint(point)
+
+    def containsObject(self, obj):
+        return self.footprint.containsObject(obj)
+
+    def uniformPointInner(self):
+        triangle, bounds = random.choices(
+            self.trianglesAndBounds,
+            cum_weights=self.cumulativeTriangleAreas)[0]
+        minx, miny, maxx, maxy = bounds
+        # TODO improve?
+        while True:
+            x, y = random.uniform(minx, maxx), random.uniform(miny, maxy)
+            if triangle.intersects(shapely.geometry.Point(x, y)):
+                return self.orient(Vector(x, y, self.z))
+
+    def intersects(self, other, triedReversed=False):
+        if isinstance(other, PolygonalRegion):
+            if self.z != other.z:
```

Comment:
Will fail if `self.z` is random. But we don't pass it to `Samplable.__init__` either, so maybe you're disallowing that? In which case we'd better have a test in the initializer and explain the restriction in the docstring.